packages/dashboard/src/customization/widgets/lineSymbol/lineAnchor/component.tsx

```diff
@@ -0,0 +1,96 @@
+import React, { CSSProperties, useEffect, useRef } from 'react';
+import { LineWidget } from '~/customization/widgets/types';
+import { UpdateWidgetsAction } from '~/store/actions';
+import { XYCoord, useDrag } from 'react-dnd';
+import { DashboardWidget } from '~/types';
+import { useGridSettings } from '~/components/actions/useGridSettings';
+import './component.css';
+
+export const LineAnchor: React.FC<{
+  style: CSSProperties;
+  anchorType: 'start' | 'end';
+  widget: LineWidget;
+  updateWidget: (widget: DashboardWidget) => UpdateWidgetsAction;
+}> = ({ style, anchorType, widget, updateWidget }) => {
+  // These refs are needed because we want to remember the coordinates at the start of the drag event.
+  // We don't want to recalculate the coordinates of the points on every render, only when the drag event completes.
+  const initialStartPointRef = useRef({
+    x: 0,
+    y: 0,
+  });
+  const initialEndPointRef = useRef({
+    x: 0,
+    y: 0,
+  });
+
+  const { cellSize } = useGridSettings();
+
+  function computeNewPosition(initialPosition: XYCoord, offsetDifference: XYCoord) {
+    const { x: initialX, y: initialY } = initialPosition;
+    const { x: offsetX, y: offsetY } = offsetDifference;
+
+    const widthPixels = Math.round(widget.width * cellSize);
+    const heightPixels = Math.round(widget.height * cellSize);
+
+    const newX = Math.max(0, Math.min(initialX + offsetX, widthPixels));
+    const newY = Math.max(0, Math.min(initialY + offsetY, heightPixels));
+
+    return { x: newX, y: newY } as XYCoord;
```

Comment:
is there a way to write this so we don't need to cast?

Reply:
`return { x: newX, y: newY };`